Auto-IP function with mac-ip MAP integrated in supernode

Thanks for all your efforts! I like n2n very much, it helps me and my friends a lot with NAT issues.
About the DHCP function within sn, we have pain during using it. The embedded dhcp server is very good, but lacking with function like mac-ip map.
Here is the use case:
We have some devices to get ip from sn, it is easy and good to use. Then I  connect them for some purpose with edge ip, it is working well as a start, suddenly the power lost, then all devices start to power on and get ip from sn again, this time, they did not get the original ip address, the topology is messed up, because all ip changed.

I used openvpn before, but I need p2p connection, so I use n2n. when I use openvpn, it provide a cache file to store the mac-ip map in a text file, which can make sure the old client and get the corresponding IP, no matter it is online or offline, it is fixed.

Can n2n provide such features? make sure the sn assgine the ip based on a ip-mac map if the record is there. and we can modify the text cache file to remove specific items if I do not want it keep old ip in the map.

Thanks again for the dev team, you are awesome!

_Originally posted by @cholley in https://github.com/ntop/n2n/issues/355#issuecomment-751647074_

> Could you please try to apply this [patch](https://github.com/ntop/n2n/files/5751288/patch.txt) to the supernode code, recompile it, and check if it solves your issue, and let me know?
> 
> It bases the auto-IP address generation on a hash value of the computer description, i.e. hostname in most cases (or whatever you supply with `-I < description>` edge cli parameter. We cannot use the MAC address for hashing because it might not be fully known (and set) at the time of auto IP query (virtual interface with definitve MAC address not up yet at that point in time).
> 
> Please do not rely on the generated IP addresses too strongly yet because the hashing function will change somewhat soon and the generated addresses with it.

Hi @Logan007 
I tried to apply the patch to stable 2.8 version, but it is failed.
Then I try to apply it to dev master, it is failed again:

```
root@VM-8-16-debian:~/app/n2n-dev# ./configure CFLAGS="-O3 -march=native"
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for pcap_open_live in -lpcap... no
checking for pcap_set_immediate_mode in -lpcap... no
checking for cap_get_proc in -lcap... no
configure: creating ./config.status
config.status: creating Makefile
config.status: creating tools/Makefile
config.status: creating include/config.h
root@VM-8-16-debian:~/app/n2n-dev# 
root@VM-8-16-debian:~/app/n2n-dev# 
root@VM-8-16-debian:~/app/n2n-dev# make
/bin/sh: 1: pkg-config: not found
gcc -O3 -march=native -I ./include        -c -o src/transform_aes.o src/transform_aes.c
gcc -O3 -march=native -I ./include        -c -o src/tuntap_osx.o src/tuntap_osx.c
gcc -O3 -march=native -I ./include        -c -o src/random_numbers.o src/random_numbers.c
src/random_numbers.c: In function ‘n2n_seed’:
src/random_numbers.c:146:12: warning: implicit declaration of function ‘_rdseed64_step’; did you mean ‘_rdrand64_step’? [-Wimplicit-function-declaration]
         if(_rdseed64_step((unsigned long long*)&seed)) {
            ^~~~~~~~~~~~~~
            _rdrand64_step
gcc -O3 -march=native -I ./include        -c -o src/speck.o src/speck.c
gcc -O3 -march=native -I ./include        -c -o src/sn.o src/sn.c
gcc -O3 -march=native -I ./include        -c -o src/example_edge_embed.o src/example_edge_embed.c
gcc -O3 -march=native -I ./include        -c -o src/sn_utils.o src/sn_utils.c
gcc -O3 -march=native -I ./include        -c -o src/network_traffic_filter.o src/network_traffic_filter.c
gcc -O3 -march=native -I ./include        -c -o src/minilzo.o src/minilzo.c
gcc -O3 -march=native -I ./include        -c -o src/edge_utils_win32.o src/edge_utils_win32.c
gcc -O3 -march=native -I ./include        -c -o src/example_sn_embed.o src/example_sn_embed.c
gcc -O3 -march=native -I ./include        -c -o src/edge.o src/edge.c
gcc -O3 -march=native -I ./include        -c -o src/transform_cc20.o src/transform_cc20.c
gcc -O3 -march=native -I ./include        -c -o src/tuntap_linux.o src/tuntap_linux.c
gcc -O3 -march=native -I ./include        -c -o src/wire.o src/wire.c
gcc -O3 -march=native -I ./include        -c -o src/transform_null.o src/transform_null.c
gcc -O3 -march=native -I ./include        -c -o src/pearson.o src/pearson.c
gcc -O3 -march=native -I ./include        -c -o src/example_edge_embed_quick_edge_init.o src/example_edge_embed_quick_edge_init.c
gcc -O3 -march=native -I ./include        -c -o src/cc20.o src/cc20.c
gcc -O3 -march=native -I ./include        -c -o src/n2n_regex.o src/n2n_regex.c
gcc -O3 -march=native -I ./include        -c -o src/tf.o src/tf.c
gcc -O3 -march=native -I ./include        -c -o src/tuntap_netbsd.o src/tuntap_netbsd.c
gcc -O3 -march=native -I ./include        -c -o src/aes.o src/aes.c
gcc -O3 -march=native -I ./include        -c -o src/header_encryption.o src/header_encryption.c
gcc -O3 -march=native -I ./include        -c -o src/edge_utils.o src/edge_utils.c
gcc -O3 -march=native -I ./include        -c -o src/transform_tf.o src/transform_tf.c
gcc -O3 -march=native -I ./include        -c -o src/n2n.o src/n2n.c
gcc -O3 -march=native -I ./include        -c -o src/tuntap_freebsd.o src/tuntap_freebsd.c
gcc -O3 -march=native -I ./include        -c -o src/transform_speck.o src/transform_speck.c
gcc -O3 -march=native -I ./include        -c -o src/sn_selection.o src/sn_selection.c
ar rcs libn2n.a  src/example_sn_embed.o  src/tuntap_osx.o  src/example_edge_embed.o  src/sn.o  src/network_traffic_filter.o  src/tf.o  src/sn_utils.o  src/wire.o  src/transform_aes.o  src/pearson.o  src/example_edge_embed_quick_edge_init.o  src/tuntap_netbsd.o  src/edge_utils.o  src/n2n.o  src/tuntap_freebsd.o  src/random_numbers.o  src/speck.o  src/minilzo.o  src/transform_tf.o  src/cc20.o  src/edge_utils_win32.o  src/edge.o  src/transform_cc20.o  src/tuntap_linux.o  src/n2n_regex.o  src/transform_null.o  src/aes.o  src/header_encryption.o  src/transform_speck.o  src/sn_selection.o
gcc -O3 -march=native -I ./include      src/edge.c  libn2n.a  -o edge
/usr/bin/ld: libn2n.a(random_numbers.o): in function `n2n_seed':
random_numbers.c:(.text+0x1c7): undefined reference to `_rdseed64_step'
collect2: error: ld returned 1 exit status
make: *** [Makefile:75: edge] Error 1

```
If possible, can you create a patch for stable version 2.8, it is better for  production  use.

> I do not understand why the compiler complains about the random part which is not involved in this patch. By the way, I would recommend to perform a `make clean` after `./configure CFLAGS="..."` and before `make`.
> 
> Anyways, if you dare to manually modify the code, you can apply the following changes to the `src/sn_utils.c` file (it will **not** work with 2.8 because the description field has just recently been added):
> 
> 1. Find and replace the `int assign_one_ip_addr()` function with this code (including the function signature, one more parameter):
> 
> ```c
> /** The IP address assigned to the edge by the auto ip address function of sn. */
> static int assign_one_ip_addr (struct sn_community *comm, n2n_desc_t dev_desc,
>                                n2n_ip_subnet_t *ip_addr) {
> 
>   struct peer_info *peer, *tmp_peer;
>   uint32_t tmp, success, net_id, mask, max_host, host_id = 1;
>   dec_ip_bit_str_t ip_bit_str = {'\0'};
> 
>   mask = bitlen2mask(comm->auto_ip_net.net_bitlen);
>   net_id = comm->auto_ip_net.net_addr & mask;
>   max_host = ~mask;
> 
>   // sorting is a prerequisite for more efficient availabilitiy check
>   HASH_SORT(comm->edges, peer_tap_ip_sort);
> 
>   // first proposal derived from hash of mac address
>   tmp = pearson_hash_32(dev_desc, sizeof(n2n_desc_t)) & max_host;
>   if(tmp == 0)        tmp++; /* avoid 0 host */
>   if(tmp == max_host) tmp--; /* avoid broadcast address */
>   tmp |= net_id;
> 
>   // candidate
>   ip_addr->net_bitlen = comm->auto_ip_net.net_bitlen;
> 
>   // check for availability starting from proposal, then downwards, ...
>   for(host_id = tmp; host_id > net_id; host_id--) {
>     ip_addr->net_addr = host_id;
>     success = ip_addr_available(comm, ip_addr);
>     if(success) {
>       break;
>     }
>   }
>   // ... then upwards
>   if(!success) {
>     for(host_id = tmp + 1; host_id < (net_id + max_host); host_id++) {
>       ip_addr->net_addr = host_id;
>       success = ip_addr_available(comm, ip_addr);
>       if(success) {
>         break;
>       }
>     }
>   }
> 
>   if(success) {
>     traceEvent(TRACE_INFO, "Assign IP %s to tap adapter of edge.", ip_subnet_to_str(ip_bit_str, ip_addr));
>     return 0;
>   } else {
>      traceEvent(TRACE_WARNING, "No assignable IP to edge tap adapter.");
>      return -1;
>   }
> }
> ```
> 
> 1. Above this function, insert a new function:
> 
> ```c
> /** checks if a certain ip address is still available, i.e. not used by any other edge of a given community */
> static int ip_addr_available (struct sn_community *comm, n2n_ip_subnet_t *ip_addr) {
> 
>     int success = 1;
>     struct peer_info *peer, *tmp_peer;
> 
>     // prerequisite: list of peers is sorted according to peer's tap ip address
>     HASH_ITER(hh, comm->edges, peer, tmp_peer) {
>         if(peer->dev_addr.net_addr  > ip_addr->net_addr) {
>             break;
>         }
>         if(peer->dev_addr.net_addr == ip_addr->net_addr) {
>             success = 0;
>             break;
>         }
>     }
>     return success;
> }
> ```
> 
> 1. Find and change the call of `assign_one_ip_addr()`, there is only one, and change it to
> 
> ```c
> assign_one_ip_addr(comm, reg.dev_desc, &ipaddr);
> ```

Hi @Logan007 
I made a clean compile without the patch, same result:
```
root@VM-8-16-debian:~/app/n2n-dev# ./autogen.sh 
Wait please...
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for pcap_open_live in -lpcap... no
checking for pcap_set_immediate_mode in -lpcap... no
checking for cap_get_proc in -lcap... no
configure: creating ./config.status
config.status: creating Makefile
config.status: creating tools/Makefile
config.status: creating include/config.h
root@VM-8-16-debian:~/app/n2n-dev# ./configure CFLAGS="-O3 -march=native"
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for pcap_open_live in -lpcap... no
checking for pcap_set_immediate_mode in -lpcap... no
checking for cap_get_proc in -lcap... no
configure: creating ./config.status
config.status: creating Makefile
config.status: creating tools/Makefile
config.status: creating include/config.h
config.status: include/config.h is unchanged
root@VM-8-16-debian:~/app/n2n-dev# make
/bin/sh: 1: pkg-config: not found
gcc -O3 -march=native -I ./include        -c -o src/transform_aes.o src/transform_aes.c
gcc -O3 -march=native -I ./include        -c -o src/tuntap_osx.o src/tuntap_osx.c
gcc -O3 -march=native -I ./include        -c -o src/random_numbers.o src/random_numbers.c
src/random_numbers.c: In function ‘n2n_seed’:
src/random_numbers.c:146:12: warning: implicit declaration of function ‘_rdseed64_step’; did you mean ‘_rdrand64_step’? [-Wimplicit-function-declaration]
         if(_rdseed64_step((unsigned long long*)&seed)) {
            ^~~~~~~~~~~~~~
            _rdrand64_step
gcc -O3 -march=native -I ./include        -c -o src/speck.o src/speck.c
gcc -O3 -march=native -I ./include        -c -o src/sn.o src/sn.c
gcc -O3 -march=native -I ./include        -c -o src/example_edge_embed.o src/example_edge_embed.c
gcc -O3 -march=native -I ./include        -c -o src/sn_utils.o src/sn_utils.c
gcc -O3 -march=native -I ./include        -c -o src/network_traffic_filter.o src/network_traffic_filter.c
gcc -O3 -march=native -I ./include        -c -o src/minilzo.o src/minilzo.c
gcc -O3 -march=native -I ./include        -c -o src/edge_utils_win32.o src/edge_utils_win32.c
gcc -O3 -march=native -I ./include        -c -o src/example_sn_embed.o src/example_sn_embed.c
gcc -O3 -march=native -I ./include        -c -o src/edge.o src/edge.c
gcc -O3 -march=native -I ./include        -c -o src/transform_cc20.o src/transform_cc20.c
gcc -O3 -march=native -I ./include        -c -o src/tuntap_linux.o src/tuntap_linux.c
gcc -O3 -march=native -I ./include        -c -o src/wire.o src/wire.c
gcc -O3 -march=native -I ./include        -c -o src/transform_null.o src/transform_null.c
gcc -O3 -march=native -I ./include        -c -o src/pearson.o src/pearson.c
gcc -O3 -march=native -I ./include        -c -o src/example_edge_embed_quick_edge_init.o src/example_edge_embed_quick_edge_init.c
gcc -O3 -march=native -I ./include        -c -o src/cc20.o src/cc20.c
gcc -O3 -march=native -I ./include        -c -o src/n2n_regex.o src/n2n_regex.c
gcc -O3 -march=native -I ./include        -c -o src/tf.o src/tf.c
gcc -O3 -march=native -I ./include        -c -o src/tuntap_netbsd.o src/tuntap_netbsd.c
gcc -O3 -march=native -I ./include        -c -o src/aes.o src/aes.c
gcc -O3 -march=native -I ./include        -c -o src/header_encryption.o src/header_encryption.c
gcc -O3 -march=native -I ./include        -c -o src/edge_utils.o src/edge_utils.c
gcc -O3 -march=native -I ./include        -c -o src/transform_tf.o src/transform_tf.c
gcc -O3 -march=native -I ./include        -c -o src/n2n.o src/n2n.c
gcc -O3 -march=native -I ./include        -c -o src/tuntap_freebsd.o src/tuntap_freebsd.c
gcc -O3 -march=native -I ./include        -c -o src/transform_speck.o src/transform_speck.c
gcc -O3 -march=native -I ./include        -c -o src/sn_selection.o src/sn_selection.c
ar rcs libn2n.a  src/example_sn_embed.o  src/tuntap_osx.o  src/example_edge_embed.o  src/sn.o  src/network_traffic_filter.o  src/tf.o  src/sn_utils.o  src/wire.o  src/transform_aes.o  src/pearson.o  src/example_edge_embed_quick_edge_init.o  src/tuntap_netbsd.o  src/edge_utils.o  src/n2n.o  src/tuntap_freebsd.o  src/random_numbers.o  src/speck.o  src/minilzo.o  src/transform_tf.o  src/cc20.o  src/edge_utils_win32.o  src/edge.o  src/transform_cc20.o  src/tuntap_linux.o  src/n2n_regex.o  src/transform_null.o  src/aes.o  src/header_encryption.o  src/transform_speck.o  src/sn_selection.o
gcc -O3 -march=native -I ./include      src/edge.c  libn2n.a  -o edge
/usr/bin/ld: libn2n.a(random_numbers.o): in function `n2n_seed':
random_numbers.c:(.text+0x1c7): undefined reference to `_rdseed64_step'
collect2: error: ld returned 1 exit status
make: *** [Makefile:75: edge] Error 1
root@VM-8-16-debian:~/app/n2n-dev# 
```
So it might be dev code issues.

> Could you please try once again but without `-march=native`?

```
root@VM-8-16-debian:~/app/n2n-dev# make clean
/bin/sh: 1: pkg-config: not found
rm -rf  src/random_numbers.o  src/transform_aes.o  src/tuntap_osx.o  src/speck.o  src/sn.o  src/example_edge_embed.o  src/sn_utils.o  src/network_traffic_filter.o  src/minilzo.o  src/transform_tf.o  src/edge_utils_win32.o  src/example_sn_embed.o  src/edge.o  src/transform_cc20.o  src/tuntap_linux.o  src/tuntap_netbsd.o  src/n2n_regex.o  src/transform_null.o  src/pearson.o  src/example_edge_embed_quick_edge_init.o  src/cc20.o  src/tf.o  src/wire.o  src/aes.o  src/header_encryption.o  src/edge_utils.o  src/n2n.o  src/tuntap_freebsd.o  src/transform_speck.o  src/sn_selection.o libn2n.a edge supernode example_edge_embed_quick_edge_init example_edge_embed example_sn_embed edge.8.gz supernode.1.gz n2n.7.gz test n2n-decode *.dSYM *~
make -C tools clean
make[1]: Entering directory '/root/app/n2n-dev/tools'
rm -rf n2n-benchmark  *.o *.dSYM *~
make[1]: Leaving directory '/root/app/n2n-dev/tools'
root@VM-8-16-debian:~/app/n2n-dev# ./configure 
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for pcap_open_live in -lpcap... no
checking for pcap_set_immediate_mode in -lpcap... no
checking for cap_get_proc in -lcap... no
configure: creating ./config.status
config.status: creating Makefile
config.status: creating tools/Makefile
config.status: creating include/config.h
config.status: include/config.h is unchanged
root@VM-8-16-debian:~/app/n2n-dev# make
/bin/sh: 1: pkg-config: not found
gcc -g -O2 -I ./include        -c -o src/transform_aes.o src/transform_aes.c
gcc -g -O2 -I ./include        -c -o src/tuntap_osx.o src/tuntap_osx.c
gcc -g -O2 -I ./include        -c -o src/random_numbers.o src/random_numbers.c
gcc -g -O2 -I ./include        -c -o src/speck.o src/speck.c
gcc -g -O2 -I ./include        -c -o src/sn.o src/sn.c
gcc -g -O2 -I ./include        -c -o src/example_edge_embed.o src/example_edge_embed.c
gcc -g -O2 -I ./include        -c -o src/sn_utils.o src/sn_utils.c
gcc -g -O2 -I ./include        -c -o src/network_traffic_filter.o src/network_traffic_filter.c
gcc -g -O2 -I ./include        -c -o src/minilzo.o src/minilzo.c
gcc -g -O2 -I ./include        -c -o src/edge_utils_win32.o src/edge_utils_win32.c
gcc -g -O2 -I ./include        -c -o src/example_sn_embed.o src/example_sn_embed.c
gcc -g -O2 -I ./include        -c -o src/edge.o src/edge.c
gcc -g -O2 -I ./include        -c -o src/transform_cc20.o src/transform_cc20.c
gcc -g -O2 -I ./include        -c -o src/tuntap_linux.o src/tuntap_linux.c
gcc -g -O2 -I ./include        -c -o src/wire.o src/wire.c
gcc -g -O2 -I ./include        -c -o src/transform_null.o src/transform_null.c
gcc -g -O2 -I ./include        -c -o src/pearson.o src/pearson.c
gcc -g -O2 -I ./include        -c -o src/example_edge_embed_quick_edge_init.o src/example_edge_embed_quick_edge_init.c
gcc -g -O2 -I ./include        -c -o src/cc20.o src/cc20.c
gcc -g -O2 -I ./include        -c -o src/n2n_regex.o src/n2n_regex.c
gcc -g -O2 -I ./include        -c -o src/tf.o src/tf.c
gcc -g -O2 -I ./include        -c -o src/tuntap_netbsd.o src/tuntap_netbsd.c
gcc -g -O2 -I ./include        -c -o src/aes.o src/aes.c
gcc -g -O2 -I ./include        -c -o src/header_encryption.o src/header_encryption.c
gcc -g -O2 -I ./include        -c -o src/edge_utils.o src/edge_utils.c
gcc -g -O2 -I ./include        -c -o src/transform_tf.o src/transform_tf.c
gcc -g -O2 -I ./include        -c -o src/n2n.o src/n2n.c
gcc -g -O2 -I ./include        -c -o src/tuntap_freebsd.o src/tuntap_freebsd.c
gcc -g -O2 -I ./include        -c -o src/transform_speck.o src/transform_speck.c
gcc -g -O2 -I ./include        -c -o src/sn_selection.o src/sn_selection.c
ar rcs libn2n.a  src/example_sn_embed.o  src/tuntap_osx.o  src/example_edge_embed.o  src/sn.o  src/network_traffic_filter.o  src/tf.o  src/sn_utils.o  src/wire.o  src/transform_aes.o  src/pearson.o  src/example_edge_embed_quick_edge_init.o  src/tuntap_netbsd.o  src/edge_utils.o  src/n2n.o  src/tuntap_freebsd.o  src/random_numbers.o  src/speck.o  src/minilzo.o  src/transform_tf.o  src/cc20.o  src/edge_utils_win32.o  src/edge.o  src/transform_cc20.o  src/tuntap_linux.o  src/n2n_regex.o  src/transform_null.o  src/aes.o  src/header_encryption.o  src/transform_speck.o  src/sn_selection.o
gcc -g -O2 -I ./include      src/edge.c  libn2n.a  -o edge
gcc -g -O2 -I ./include      src/sn.c  libn2n.a  -o supernode
/usr/bin/ld: libn2n.a(sn_utils.o): in function `process_udp':
/root/app/n2n-dev/src/sn_utils.c:1294: undefined reference to `n2n_rand_sqr'
collect2: error: ld returned 1 exit status
make: *** [Makefile:78: supernode] Error 1
root@VM-8-16-debian:~/app/n2n-dev# 

```

--- a/src/sn_utils.c
+++ b/src/sn_utils.c
@@ -455,6 +455,27 @@ static int update_edge (n2n_sn_t *sss,
 }
 
 
+/** checks if a certain ip address is still available, i.e. not used by any other edge of a given community */
+static int ip_addr_available (struct sn_community *comm, n2n_ip_subnet_t *ip_addr) {
+
+    int success = 1;
+    struct peer_info *peer, *tmp_peer;
+
+    // prerequisite: list of peers is sorted according to peer's tap ip address
+    HASH_ITER(hh, comm->edges, peer, tmp_peer) {
+        if(peer->dev_addr.net_addr  > ip_addr->net_addr) {
+            break;
+        }
+        if(peer->dev_addr.net_addr == ip_addr->net_addr) {
+            success = 0;
+            break;
+        }
+    }
+
+    return success;
+}
+
+
 static signed int peer_tap_ip_sort (struct peer_info *a, struct peer_info *b) {
 
     uint32_t a_host_id = a->dev_addr.net_addr & (~bitlen2mask(a->dev_addr.net_bitlen));
@@ -465,40 +486,54 @@ static signed int peer_tap_ip_sort (struct peer_info *a, struct peer_info *b) {
 
 
 /** The IP address assigned to the edge by the auto ip address function of sn. */
-static int assign_one_ip_addr (struct sn_community *comm,
-                               n2n_ip_subnet_t *ipaddr) {
+static int assign_one_ip_addr (struct sn_community *comm, n2n_desc_t dev_desc, n2n_ip_subnet_t *ip_addr) {
 
-    struct peer_info *peer, *tmpPeer;
-    uint32_t net_id, mask, max_host, host_id = 1;
+    struct peer_info *peer, *tmp_peer;
+    uint32_t tmp, success, net_id, mask, max_host, host_id = 1;
     dec_ip_bit_str_t ip_bit_str = {'\0'};
 
     mask = bitlen2mask(comm->auto_ip_net.net_bitlen);
     net_id = comm->auto_ip_net.net_addr & mask;
     max_host = ~mask;
 
+    // sorting is a prerequisite for more efficient availabilitiy check
     HASH_SORT(comm->edges, peer_tap_ip_sort);
-    HASH_ITER(hh, comm->edges, peer, tmpPeer) {
-        if((peer->dev_addr.net_addr & bitlen2mask(peer->dev_addr.net_bitlen)) == net_id) {
-            if(host_id >= max_host) {
-                traceEvent(TRACE_WARNING, "No assignable IP to edge tap adapter.");
-                return -1;
-            }
-            if(peer->dev_addr.net_addr == 0) {
-                continue;
-            }
-            if((peer->dev_addr.net_addr & max_host) == host_id) {
-                ++host_id;
-            } else {
+
+    // first proposal derived from hash of mac address
+    tmp = pearson_hash_32(dev_desc, sizeof(n2n_desc_t)) & max_host;
+    if(tmp == 0)        tmp++; /* avoid 0 host */
+    if(tmp == max_host) tmp--; /* avoid broadcast address */
+    tmp |= net_id;
+
+    // candidate
+    ip_addr->net_bitlen = comm->auto_ip_net.net_bitlen;
+
+    // check for availability starting from proposal, then downwards, ...
+    for(host_id = tmp; host_id > net_id; host_id--) {
+        ip_addr->net_addr = host_id;
+        success = ip_addr_available(comm, ip_addr);
+        if(success) {
+            break;
+        }
+    }
+    // ... then upwards
+    if(!success) {
+        for(host_id = tmp + 1; host_id < (net_id + max_host); host_id++) {
+            ip_addr->net_addr = host_id;
+            success = ip_addr_available(comm, ip_addr);
+            if(success) {
                 break;
             }
         }
     }
-    ipaddr->net_addr = net_id | host_id;
-    ipaddr->net_bitlen = comm->auto_ip_net.net_bitlen;
 
-    traceEvent(TRACE_INFO, "Assign IP %s to tap adapter of edge.", ip_subnet_to_str(ip_bit_str, ipaddr));
-
-    return 0;
+    if(success) {
+        traceEvent(TRACE_INFO, "Assign IP %s to tap adapter of edge.", ip_subnet_to_str(ip_bit_str, ip_addr));
+        return 0;
+    } else {
+        traceEvent(TRACE_WARNING, "No assignable IP to edge tap adapter.");
+        return -1;
+    }
 }
 
 
@@ -1235,7 +1270,7 @@ static int process_udp (n2n_sn_t * sss,
                 if((reg.dev_addr.net_addr == 0) || (reg.dev_addr.net_addr == 0xFFFFFFFF) || (reg.dev_addr.net_bitlen == 0) ||
                    ((reg.dev_addr.net_addr & 0xFFFF0000) == 0xA9FE0000 /* 169.254.0.0 */)) {
                     memset(&ipaddr, 0, sizeof(n2n_ip_subnet_t));
-                    assign_one_ip_addr(comm, &ipaddr);
+                    assign_one_ip_addr(comm, reg.dev_desc, &ipaddr);
                     ack.dev_addr.net_addr = ipaddr.net_addr;
                     ack.dev_addr.net_bitlen = ipaddr.net_bitlen;
                 }

